File: 621.task-scheduler.cpp
/*
 * @lc app=leetcode id=621 lang=cpp
 *
 * [621] Task Scheduler
 */

// @lc code=start
class Solution {
public:
    class Task {
        public:
        int nums;
        char letter;
        Task(int a, char b) :nums(a), letter(b){}
    } ;

    // bool compare(Task i1, Task i2){
    //     return (i1.nums > i2.nums);
    // }

    int leastInterval(vector<char>& tasks, int n) {
        vector<char> result;
        vector<Task> Tasklist;
        for(int i = 0; i < 26; i++){
            Task temp_task(0,'A'+i);
            Tasklist.push_back(temp_task);
        }
        for(int i = 0; i < tasks.size(); i++){
            Tasklist[tasks[i]-'A'].nums ++;
        }
        sort(Tasklist.begin(), Tasklist.end(), [](const Task& lhs, const Task& rhs) {
        return lhs.nums > rhs.nums;
        });

        while(Tasklist[0].nums > 0){
            int rank = -1;
            int size = result.size();
            // for(int i = 0; i < Tasklist.size(); i++){
            //     if(Tasklist[i].nums>0)cout<<Tasklist[i].letter<<" "<<Tasklist[i].nums<<" ";
            // }
            // cout <<endl;
            for(int i = 0; i < Tasklist.size(); i++){
                rank = i;
                for(int j = 1; j <= n&& j<=size; j ++){
                    if(Tasklist[i].letter  == result[size - j]){
                        rank = -1;
                        break;
                    }
                }
                if(rank >= 0)break;
            }
            if(rank < 0 ){
                result.push_back('+');
                continue;
            }
            if(Tasklist[rank].nums>0){
                Tasklist[rank].nums--;
                result.push_back(Tasklist[rank].letter);
            }
            else{
                result.push_back('+');
                continue;
            }
            if(rank >=0 &&rank < Tasklist.size()-1){
                int swap_pos = rank;
                int end = Tasklist.size();
                int mid = 0;
                while(swap_pos < end){
                    mid = swap_pos + (end - swap_pos)/2;
                    if(Tasklist[rank].nums < Tasklist[mid].nums){
                        swap_pos = mid + 1;
                    }
                    else end = mid;
                }
                // cout<< "rank = "<< rank << " " << Tasklist[rank].letter <<" mid =  "<<end<< " "<< Tasklist[mid].letter<<endl;
                // for(;swap_pos<Tasklist.size();swap_pos++){
                //     if(Tasklist[rank].nums >= Tasklist[swap_pos].nums)break;
                // }
                if(swap_pos>rank){
                    int temp_int = Tasklist[rank].nums;
                    char temp_char = Tasklist[rank].letter;
                    Tasklist[rank].nums = Tasklist[swap_pos-1].nums;
                    Tasklist[rank].letter = Tasklist[swap_pos-1].letter;
                    Tasklist[swap_pos-1].nums = temp_int;
                    Tasklist[swap_pos-1].letter = temp_char;
                }
            }
        }
        // for(int i = 0; i < result.size(); i++){
        //     cout<<result[i]<<" ";
        // }
        return result.size();

        
    }
};
// @lc code=end

File: 621.task-scheduler.py
#
# @lc app=leetcode id=621 lang=python3
#
# [621] Task Scheduler
#
# https://leetcode.com/problems/task-scheduler/description/
#
# algorithms
# Medium (56.20%)
# Likes:    7990
# Dislikes: 1592
# Total Accepted:    414.1K
# Total Submissions: 736.7K
# Testcase Example:  '["A","A","A","B","B","B"]\n2'
#
# Given a characters array tasks, representing the tasks a CPU needs to do,
# where each letter represents a different task. Tasks could be done in any
# order. Each task is done in one unit of time. For each unit of time, the CPU
# could complete either one task or just be idle.
# 
# However, there is a non-negative integer n that represents the cooldown
# period between two same tasks (the same letter in the array), that is that
# there must be at least n units of time between any two same tasks.
# 
# Return the least number of units of times that the CPU will take to finish
# all the given tasks.
# 
# 
# Example 1:
# 
# 
# Input: tasks = ["A","A","A","B","B","B"], n = 2
# Output: 8
# Explanation: 
# A -> B -> idle -> A -> B -> idle -> A -> B
# There is at least 2 units of time between any two same tasks.
# 
# 
# Example 2:
# 
# 
# Input: tasks = ["A","A","A","B","B","B"], n = 0
# Output: 6
# Explanation: On this case any permutation of size 6 would work since n = 0.
# ["A","A","A","B","B","B"]
# ["A","B","A","B","A","B"]
# ["B","B","B","A","A","A"]
# ...
# And so on.
# 
# 
# Example 3:
# 
# 
# Input: tasks = ["A","A","A","A","A","A","B","C","D","E","F","G"], n = 2
# Output: 16
# Explanation: 
# One possible solution is
# A -> B -> C -> A -> D -> E -> A -> F -> G -> A -> idle -> idle -> A -> idle
# -> idle -> A
# 
# 
# 
# Constraints:
# 
# 
# 1 <= task.length <= 10^4
# tasks[i] is upper-case English letter.
# The integer n is in the range [0, 100].
# 
# 
#

# @lc code=start
import heapq
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        if(not n):return len(tasks)
        count = {}
        for letter in tasks:
            if letter in count:
                count[letter] += 1
            else:
                count[letter] = 1
        heap = []
        result = []
        time = 0
        for key in count:
            heapq.heappush(heap,-count[key])
        n += 1
        offset = 0
        tomin = 0 
        # print(heap)
        while(heap):
            
            # print(result)
            while(len(result) < n and heap):
                num = heapq.heappop(heap)
                result.append(-num)
            # if(heap):tomin = -heap[0] - 1
            # else : tomin = result[-1]
            tomin = 1
            time += tomin * n
            offset = n - len(result)
            for temp in result:
                if(tomin  >= temp):break
                heapq.heappush(heap,tomin - temp)
            result.clear()
            # print(heap)
            # print(time)
        time = time - offset
        return time
        



# @lc code=end

